File: backend/app/api/pipeline.py
"""Pipeline 视频生成流水线 - API 路由

一键生成：串联 M1→M4→M2→M5→M6→M8 的完整视频生成流程。
"""

from __future__ import annotations
from fastapi import APIRouter, HTTPException

from app.models.pipeline import (
    PipelineGenerateRequest, PipelineGenerateResponse,
    PipelineProject, PipelineStatusResponse,
)
from app.services.pipeline.engine import pipeline_engine

router = APIRouter()


@router.post("/generate", response_model=PipelineGenerateResponse)
async def generate_video(req: PipelineGenerateRequest):
    """一键生成视频 - 执行完整流水线

    流程:
    1. M1 提示词优化 → 生成标准化绘图关键词
    2. M4 文案生成 → 生成结构化视频文案
    3. M2 AI绘图 → 为每个段落生成配图
    4. M5 TTS配音 → 为每个段落合成语音
    5. M6 智能剪辑 → 图片+音频+字幕合成为视频
    6. M8 质量评估 → 综合评分和建议
    """
    if not req.topic.strip():
        raise HTTPException(status_code=400, detail="视频主题不能为空")
    return pipeline_engine.generate(req)


@router.get("/projects", response_model=list[PipelineProject])
async def list_projects():
    """获取所有 Pipeline 项目"""
    return pipeline_engine.list_projects()


@router.get("/projects/{project_id}", response_model=PipelineProject)
async def get_project(project_id: str):
    """查询单个 Pipeline 项目详情"""
    project = pipeline_engine.get_project(project_id)
    if not project:
        raise HTTPException(status_code=404, detail="项目不存在")
    return project


@router.get("/projects/{project_id}/status", response_model=PipelineStatusResponse)
async def get_project_status(project_id: str):
    """查询 Pipeline 项目状态"""
    project = pipeline_engine.get_project(project_id)
    if not project:
        raise HTTPException(status_code=404, detail="项目不存在")

    result_url = None
    for stage in project.stages:
        if stage.stage.value == "edit" and stage.status.value == "completed":
            result_url = stage.output.get("video_url")
            break

    return PipelineStatusResponse(
        project_id=project.id,
        status=project.status,
        stages=project.stages,
        overall_progress=project.overall_progress,
        result_video_url=result_url,
    )

File: frontend/src/api/pipeline.ts
/** Pipeline 视频生成 - API 服务层 */

const BASE_URL = 'http://localhost:8000/api/v1/pipeline'

async function request<T>(url: string, options?: RequestInit): Promise<T> {
  const res = await fetch(url, {
    headers: { 'Content-Type': 'application/json' },
    ...options,
  })
  if (!res.ok) {
    const err = await res.json().catch(() => ({ detail: res.statusText }))
    throw new Error(err.detail || `HTTP ${res.status}`)
  }
  return res.json()
}

// ---- 类型定义 ----

export interface PipelineStageResult {
  stage: string
  status: 'pending' | 'running' | 'completed' | 'failed' | 'skipped'
  progress: number
  message: string
  output: Record<string, any>
  error?: string
}

export interface PipelineGenerateRequest {
  topic: string
  genre?: string
  target_length?: string
  style?: string
  lang?: string
  voice_style?: string
  mood_tags?: string[]
  auto_generate?: boolean
}

export interface PipelineGenerateResponse {
  project_id: string
  title: string
  status: string
  stages: PipelineStageResult[]
  current_stage: string
  overall_progress: number
  estimated_duration_sec: number
  created_at: string
}

export interface PipelineProject {
  id: string
  title: string
  topic: string
  genre: string
  status: string
  stages: PipelineStageResult[]
  overall_progress: number
  created_at: string
}

// ---- API 函数 ----

/** 一键生成视频 - 执行完整流水线 */
export async function generateVideo(req: PipelineGenerateRequest): Promise<PipelineGenerateResponse> {
  return request<PipelineGenerateResponse>(`${BASE_URL}/generate`, {
    method: 'POST',
    body: JSON.stringify(req),
  })
}

/** 获取所有项目 */
export async function listProjects(): Promise<PipelineProject[]> {
  return request<PipelineProject[]>(`${BASE_URL}/projects`)
}

/** 获取项目详情 */
export async function getProject(projectId: string): Promise<PipelineProject> {
  return request<PipelineProject>(`${BASE_URL}/projects/${projectId}`)
}

/** 查询项目状态 */
export async function getProjectStatus(projectId: string): Promise<{
  project_id: string
  status: string
  stages: PipelineStageResult[]
  overall_progress: number
  result_video_url?: string
}> {
  return request(`${BASE_URL}/projects/${projectId}/status`)
}
